scripts/analytics/upload.py

```diff
@@ -0,0 +1,135 @@
+            waterbutler_upload(waterbutler_path, file_name)
+
+
+def create_dir(dir_name, self_id, parent_path, parent_id):
+    assert self_id is None, 'Cannot create directories that already exist'
+    path = '/'
+    if parent_path is not None and len(parent_path) > 0:
+        path = '/' + parent_id + '/'
+    name = dir_name
+    create = True
+
+    node = models.Node.load(settings.TABULATE_LOGS_NODE_ID)
+    user = models.User.load(settings.TABULATE_LOGS_USER_ID)
+
+    retry = 0
+    while retry < 3:  # server 503 may still happen with sleep(1), retry for 3 times
+        resp = utils.send_file(name, 'folder-upload', None, node, user, create, path)
+        sleep(1)  # walk around 503 error
+        if resp.status_code != 503:
+            break
+        retry += 1
+    if retry == 3:
+        resp.raise_for_status()
+
+
+def create_or_update_file(file_stream, file_name, self_id, parent_path, parent_id):
+    path = '/'
+    if parent_path is not None:
+        path = '/' + parent_id + '/'
+    if self_id is None:
+        name = file_name
+        create = True
+    else:
+        name = self_id
+        create = False
+        path = '/'
+
+    node = models.Node.load(settings.TABULATE_LOGS_NODE_ID)
+    user = models.User.load(settings.TABULATE_LOGS_USER_ID)
+
+    retry = 0
+    while retry < 3:  # server 503 may still happen with sleep(1), retry for 3 times
+        resp = utils.send_file(name, 'file-upload', file_stream, node, user, create, path)
+        sleep(1)  # walk around 503 error
+        if resp.status_code != 503:
+            break
+        retry += 1
+    if retry == 3:
+        print('response = {}'.format(resp.json()))
+        resp.raise_for_status()
+
+
+def waterbutler_upload(path, name, test_existence=False):
+    assert path.startswith('/'), 'Path must always starts with /'
+
+    root = database['storedfilenode'].find({'node': settings.ANALYTICS_LOGS_NODE_ID, 'parent': None})
+    if root.count() != 1:
+        print('Invalid Node ID: Cannot find the project node {}.'.format(settings.ANALYTICS_LOGS_NODE_ID))
+        return
+    parent_id = root[0]['_id']
+    parent_path = []
+
+    for a_dir in path.split('/')[1:]:
+        if a_dir == '':  # upload to the project folder
+            parent_path = None
+            break
+        dirs = database['storedfilenode'].find({'name': a_dir, 'parent': parent_id})
+        assert dirs.count() in [0,1], 'Database query failure'
+        if dirs.count() == 0:  # dir does not exsit
+            if test_existence:
+                return False
+            else:
+                create_dir(a_dir, None, parent_path, parent_id)
+                dirs = database['storedfilenode'].find({'name': a_dir, 'parent': parent_id})
+                if dirs.count() == 1:
+                    parent_id = dirs[0]['_id']
+                    parent_path.append(parent_id)
+                else:
+                    print('I/O Error: cannot find newly created directory')
```

Comment:
@chrisseto You are right, thanks. I had some tested code here. Forgot to change back to "raise Exception". Same for the code below.